docs/docs/protocol-specs/gas-and-fees/specifying-gas-fee-info.md

```diff
@@ -0,0 +1,134 @@
+---
+title: Specifying Gas & Fee Info
+---
+
+# Specifying Gas & Fee Info
+
+When users submit a `TxExecutionRequest` on the Aztec Network, they provide a `TxContext`, which holds `GasSettings` for the transaction.
+
+An abridged version of the class diagram is shown below:
+
+```mermaid
+classDiagram
+class TxExecutionRequest {
+  +TxContext txContext
+}
+
+class TxContext {
+  +GasSettings gasSettings
+}
+
+class GasSettings {
+  +Gas gasLimits
+  +Gas teardownGasLimits
+  +GasFees maxFeesPerGas
+  +Fr inclusionFee
+}
+
+class Gas {
+  +UInt32 daGas
+  +UInt32 l2Gas
+}
+
+class GasFees {
+  +Fr feePerDaGas
+  +Fr feePerL2Gas
+}
+
+TxContext --> GasSettings
+GasSettings --> Gas
+GasSettings --> GasFees
+```
+
+:::note 
+All fees are denominated in the [Fee Payment Asset (FPA)](./fee-payment-asset.md).
+:::
+
+# Gas Dimensions and Inclusion Fee
+
+Transactions are metered for their gas consumption across two dimensions:
+
+1. **Data Availability (DA) Gas**: This dimension measures data usage by the transaction, e.g. creating/spending notes, emitting logs, etc.
+2. **Layer 2 (L2) Gas**: This dimension measures computation usage of the public VM.
+
+This is similar to the gas model in Ethereum, where transaction consume gas to perform operations, and may also consume blob gas for storing data.
+
+Separately, every transaction has overhead costs associated with it, e.g. verifying its encompassing rollup proof on L1, which are captured in the `inclusionFee`, which is not tied to gas consumption on the transaction, but is specified in FPA.
+
+See the [Fee Schedule](./fee-schedule.md) for a detailed breakdown of costs associated with different actions.
+
+
+# `gasLimits` and `teardownGasLimits`
+
+Transactions can optionally have a "teardown" phase as part of their public execution, during which the "transaction fee" is available to public functions. This is useful to transactions/contracts that need to compute a "refund", e.g. contracts that facilitate [fee abstraction](./tx-setup-and-teardown.md).
+
+Because the transaction fee must be known at the time teardown is executed, transactions must effectively "prepay" for the teardown phase. Thus, the `teardownGasLimits` are portions of the `gasLimits` that are reserved for the teardown phase.
+
+For example, if a transaction has `gasLimits` of 1000 DA gas and 2000 L2 gas, and `teardownGasLimits` of 100 DA gas and 200 L2 gas, then the transaction will be able to consume 900 DA gas and 1800 L2 gas during the main execution phase, but 100 DA gas and 200 L2 gas **will be consumed** to cover the teardown phase: even if teardown does not consume that much gas, the transaction will still be charged for it; even if the transaction does not have a teardown phase, the gas will still be consumed.
```

Comment:
> even if the transaction does not have a teardown phase, the gas will still be consumed.

What if we just validate in the kernel tail that teardown gas must be zero if there's no teardown function enqueued?

Reply:
Only arguments I can produce are:
- consistency: you always lose teardown limits
- future proof: what if we want private teardowns? If we put up guardrails to stop you when you *erroneously* provide teardown limits, but we add a private teardown in the future, it will simply no longer error if you *erroneously* provide teardown limits. 

But if there's no way we add private teardown, then the argument is flimsy and I agree with you.

Reply:
I don't really agree with this approach of effectively using all of the teardown gas up front. To take your example, if my app logic uses 920 DA gas and teardown uses 50 DA gas, my transaction will revert, despite not breaching either the teardown or global gas limit.

Whilst the user gets charged for the complete teardown gas limit, it should still be correctly metered for the purpose of reverting.

Reply:
Additionally, this metered value of teardown gas is what should go into the block level gas used published to L1.

Reply:
In that case, the amount of gas that is available to teardown is the minimum between the teardown limit specified, and the global limit specified less non/revertible gas used.

A problem with this is that it removes any guarantee we could supply the FPC/dapp running teardown: as part of private setup it can no longer look at the amount the user has allocated for teardown and be confident that it will actually receive that amount, since app logic can bleed into it without its knowledge/control.

> this metered value of teardown gas is what should go into the block level gas used published to L1.

I agree with this. I will make that point more clear.

Reply:
> this metered value of teardown gas is what should go into the block level gas used published to L1

Are we sure about this? I was under the impression that teardown gas is always accounted as fully consumed, so there's no point in metering it except for the purpose of reverting.

Reply:
> A problem with this is that it removes any guarantee we could supply the FPC/dapp running teardown: as part of private setup it can no longer look at the amount the user has allocated for teardown and be confident that it will actually receive that amount, since app logic can bleed into it without its knowledge/control.

Yes, you are right. Ignore my comment. We just need to make clear that the is an implicit 'app logic gas limit'.

Reply:
>Are we sure about this? I was under the impression that teardown gas is always accounted as fully consumed, so there's no point in metering it except for the purpose of reverting.

Yes, I think so. In my mind the only real difference with teardown is that the user pays for all of the gas limit, regardless of how much is used. Apart from that, all of the usual gas accounting should apply.

Reply:
I'm going back on what I said before: I think the published/metered value should be the full teardown limit, as it is currently written, primarily because if we updated the `gas_used` to reflect the "accurate" amount, there will be a discrepancy between the inferred `transaction_fee` considering the published TxEffects and what the user actually paid.

I also think it is safe to include the full amount as part of the block gas used (with respect to 1559-style updates) because the user will have paid for the gas: they can't blow up the targets for free.

Last, considering they will always be charged the full teardown gas, they will be highly incentivized to have their provided value be extremely close to the true amount used.

Reply:
> I think the published/metered value should be the full teardown limit, as it is currently written, primarily because if we updated the gas_used to reflect the "accurate" amount, there will be a discrepancy between the inferred transaction_fee considering the published TxEffects and what the user actually paid.

You beat me to it. Agree that it'll make things easier if we post the total `gas_used` of a tx to be the sum of the "actual gas used during setup and app" plus "teardown limit". And we use the same value for both fee payment and block gas limit targets.

Reply:
Hmm, it just feels wrong to me somehow. Shouldn't the total gas used figure on the block be an accurate representation? Just to recap, if a tx reverts do the `gas_used` figures get set to the limits of the tx? Presumably yes if we are taking the maximum fee from the user.

> I also think it is safe to include the full amount as part of the block gas used (with respect to 1559-style updates) because the user will have paid for the gas: they can't blow up the targets for free.

But a sequencer can though can't they? They can maniuplate the block's `gas_used` to be whatever they want (below the limit) without having to do the work or incur the corresponding DA cost. Or have I misunderstood?

Reply:
Maybe we don't need to publish the `gas_used` on the block at all. If the only use for accumulating it is to check against the block limit then that will be done in the rollup circuits and we can omit it from L1.

Reply:
> But a sequencer can though can't they? They can maniuplate the block's gas_used to be whatever they want (below the limit) without having to do the work or incur the corresponding DA cost. Or have I misunderstood?

True, a sequencer could include a spurious tx with an arbitrarily high teardown gas, and fill the block with that. Whether it's costly for them or not depends on whether they keep the associated fees or not, and whether the fees are dictated by the protocol or not.

As for the actual compute cost (ie doing the work), actually running those functions or not is not a big effort, so I doubt it makes much of a difference. Eg gas limits in ethereum are tight not to protect the sequencer from doing too much work, but for ensuring that every node in the network can follow along.

But I see your point in that it's odd having a gas limit that behaves as a limit, and another gas limit that's spent automatically regardless of the work done, and having the `gas_used` not reflective of the gas actually used.

I think we may be able to make things clearer if we make `teardown_gas_limit` **not** part of the total `gas_limit`. We could rename `teardown_gas_limit` to `teardown_gas_allocation`, have `gas_used` report on the gas of app+setup only, and have the `total_gas` equal to `fixed_base_cost + gas_used + teardown_allocation`. Do you think this would help?

Reply:
> if a tx reverts do the gas_used figures get set to the limits of the tx?

Yes

> But a sequencer can though can't they?

Yes, but to no real benefit- they're charged a fee that goes back to themselves, and they're reducing the net new fees they could take in. And assuming that we adopt a max percentage change to prices from block per block, the impact on the network from a malicious sequencer that wants to increase everyone's prices should be constrained in time and size.

> Maybe we don't need to publish the gas_used on the block at all.

For sure! I think the only reason to publish `gas_used` is to allow other people to compute the transaction fees. But maybe we ought to just publish the `transaction_fee`! 

> another gas limit that's spent automatically regardless of the work done

Though `teardown_gas_limit` is still a limit in that it cannot be exceeded, it's just the case that any surplus is not refunded.

I think best is to publish `transaction_fee` instead of `gas_used` + `inclusion_fee`s, and rename `teardown_gas_limit` to `teardown_gas_allocation` to make it clear that you don't get a refund, but keep it part of `gas_limit` so you don't need to do math to get to the real limit.

Reply:
@spalladino , @PhilWindle
What do we think here? Just publish `transaction_fee` and rename `teardown_gas_limit` to `teardown_gas_allocation`?

Reply:
We can also reverse engineer the approximate breakdown by looking at the TxEffects. Given that, I agree that we should favor the user.